관리자 로그인/로그아웃, 인증/인가

hellonaeunkim opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
- 관리자 페이지 접근을 위한 로그인/로그아웃 기능 구현
- JWT 토큰 기반의 인증 시스템을 도입하여 보안성을 강화

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
1. 로그인 페이지 (Thymeleaf)
<img width="1362" alt="Image" src="https://github.com/user-attachments/assets/e3f037a3-6fb0-4b6a-9409-24060b822320" />

- JWT 토큰 기반 인증 구현
- 최초 기본 관리자 계정 자동 생성 (ID : admin, PW : admin123)
- 로그인 성공 시 관리자 페이지인 대시보드로 리다이렉트



2. 관리자 페이지 (Thymeleaf)
<img width="1362" alt="Image" src="https://github.com/user-attachments/assets/6ec0fd68-fe2e-433f-bbba-16fdd883f4e8" />

&nbsp;1) 관리자 로그아웃 기능
- 토큰 무효화 처리
- 로그아웃 시 로그인 페이지로 리다이렉트
- localStorage에서 토큰 제거

&nbsp;2) 보안 설정
- Spring Security를 통한 접근 제어
- ROLE_ADMIN 권한을 가진 사용자만 관리자 페이지 접근 가능

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->
1. PR 포인트
   1) JWT 토큰 관리
      - 토큰 생성 및 검증 로직 구현
      - 토큰 블랙리스트를 통한 로그아웃 처리
   2) 보안 설정
      - SecurityConfig를 통한 URL 별 접근 권한 설정
      - BCrypt를 사용한 비밀번호 암호화
   3) 초기 데이터 설정
      - AdminInitializerRunner를 통한 기본 관리자 계정 자동 생성

2. 궁금한 점
   - 프론트 React로 변경하는 부분에 대해 팀원분들과 상의하고 싶습니다~! 화이팅!

jurio5 committed:
Add : 관리자 페이지 구현 (타임리프) 및 관리자 컨트롤러 생성

hellonaeunkim committed:
feat : 관리자 초기 계정 정보 엔티티 생성

hellonaeunkim committed:
feat : 관리자 초기 계정 정보 엔티티 생성

jurio5 committed:
controller & entity & repository & service 구현 및 로그인 구현

jurio5 committed:
관리자 전용 로그인 대시보드 추가

hellonaeunkim committed:
fix : .gitignore 수정

hellonaeunkim committed:
fix : .gitignore 수정

hellonaeunkim committed:
refactor : AdminSetup REST API 방식으로 수정

hellonaeunkim committed:
refactor : AdminSetup REST API 방식으로 수정

jurio5 committed:
Fix : Entity

hellonaeunkim committed:
feat : 관리자 초기 계정 생성 request 추가

hellonaeunkim committed:
feat : 관리자 초기 계정 생성 request 추가

jurio5 committed:
Add : 로그아웃 기능 추가

hellonaeunkim committed:
refactor : setup 상위 package로 admin 추가

hellonaeunkim committed:
refactor : setup 상위 package로 admin 추가

hellonaeunkim committed:
feat : 관리자 초기 계정 생성 response 추가

hellonaeunkim committed:
feat : 관리자 초기 계정 생성 response 추가

jurio5 committed:
Fix : AdminEntity modify

jurio5 committed:
Fix : AdminController

hellonaeunkim committed:
feat : 관리자 회원가입 Controller, Service, Repository 생성

hellonaeunkim committed:
feat : 관리자 회원가입 Controller, Service, Repository 생성

hellonaeunkim committed:
feat : [CommandLineRunner] 관리자 계정 초기 생성 기능 구현

hellonaeunkim committed:
feat : [CommandLineRunner] 관리자 계정 초기 생성 기능 구현

hellonaeunkim committed:
style : code indent 수정, import 정리

hellonaeunkim committed:
style : code indent 수정, import 정리

dnzp75 committed:
refactor: 주문 생성 api 의 response 데이터 더 자세히 추가

- 사용자에게 자세한 정보를 나타내기 위함

dnzp75 committed:
feat: Order entity에 우편 번호 필드 추가

dnzp75 committed:
refactor: 변경사항에 따른 OrderService 수정

dnzp75 committed:
refactor: 주문 상태 한글로 반환하도록 enum 수정

dnzp75 committed:
feat: 백엔드 서버와 프론트 서버 CORS 문제 해결을 위한 관련 설정 SecurityConfig에 추가

dnzp75 committed:
test: 변경사항에 다른 테스트 코드 수정

dnzp75 committed:
feat: 서버 실행 시 임시 상품 데이터 추가

dnzp75 committed:
feat: 상품 조회 api 기능 필요함으로 인해  product 브랜치 커밋 내역 확인 후 임시로 추가

dnzp75 committed:
feat: 프론트 파일 내에 상품 주문 페이지 구현

jurio5 committed:
Add : SpringSecurity

hellonaeunkim committed:
refactor : 회원 가입 기능 삭제 & admin_password 삭제

hellonaeunkim committed:
refactor : 회원 가입 기능 삭제 & admin_password 삭제

jurio5 committed:
Fix : logout -> redirect login form

jurio5 committed:
Add : user-dashboard

jurio5 committed:
1차 폼 베이스 형식

hellonaeunkim committed:
test : 초기 관리자 생성 test 코드 통과

hellonaeunkim committed:
test : 초기 관리자 생성 test 코드 통과

jurio5 committed:
Delete : Fome Base 방식 제거 및 JWT 준비 , React 패키지 설치

jurio5 committed:
Update : gitignore to include workspace.xml

jurio5 committed:
Update : .gitignore to exclude backup gitignore files

jurio5 committed:
Add : backup gitignore files

jurio5 committed:
Force add workspace.xml despite .gitignore

jurio5 committed:
setting

jurio5 committed:
Add : JwtTokenProvider (JWT Token Create & validate
)

jurio5 committed:
Add : JwtAuthenticationFilter

jurio5 committed:
feat : Add JWT support to SecurityConfig

jurio5 committed:
feat : Add - LoginRequest & TokenResponse & AdminController Post Mapping login & AdminService Validate

jurio5 committed:
Add : JWT secret Token in yml

jurio5 committed:
Add : GlobalExceptionHandler

hellonaeunkim committed:
refactor: JWT 토큰 만료 시간 설정 파일로 분리

hellonaeunkim committed:
refactor : JWT 관련 Exception 추가

hellonaeunkim committed:
chore : reabase 충돌로 IDE 설정 업데이트

hellonaeunkim committed:
chore : IDE 설정 업데이트

hellonaeunkim committed:
feat : JWT Logout 기능 구현

jurio5 committed:
Fix: complement case

jurio5 committed:
Fix : JwtTokenProvider in Message

jurio5 committed:
Fix : H2 Connect & RESTController -> Controller & JWT 예외 개별 적용 및 의존성 순환 참조 해결

jurio5 committed:
feat : Too Many Work & 로그인 시 ROLE_ADMIN 권한을 가진 유저만 대시보드로 이동 & admin-login && dashboard JWT Token Add

jurio5 committed:
Add : JWT 관련 단일 예외 로그 처리 및 관리자 권한이 있는 유저는 모든 페이지 탐색 허용

jurio5 committed:
Fix : Log Translate

hellonaeunkim committed:
refactor : entity role 제거, password 추가

hellonaeunkim committed:
refactor : role 관련 entity, repository 파일 삭제

jurio5 committed:
Add : XSS Security CSP

hellonaeunkim committed:
chore: .gitignore 파일 업데이트

jurio5 committed:
feat : del Xss & Add SQL Injection

hellonaeunkim committed:
fix : 브랜치 병합 충돌 해결

hellonaeunkim committed:
refactor : role 관련 파일 삭제

jurio5 committed:
Clean up project and update .gitignore

jurio5 committed:
Merge branch 'feature-admin' into feature-admin-login

jurio5 committed:
feat: Admin 관련 코드 분리

jurio5 committed:
feat: Add templates and configuration files

jurio5 committed:
Fix : application

jurio5 committed:
Fix SecurityConfig

jurio5 committed:
Add : init Admin Role

jurio5 commented:
![Auth user](https://github.com/user-attachments/assets/fe0e1b1d-4a0e-429b-ac95-a1029c0658f6)
- ROLE_USER 권한으로 로그인 시도

![ban](https://github.com/user-attachments/assets/13c89059-3842-49b4-8534-39cd086c5942)
- 인가, 접속 거부

![token sha256](https://github.com/user-attachments/assets/7c33d427-1761-4b2b-b1cf-6b26bcb10949)
-  HMAC 방식으로 키 생성

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @GetMapping("/dashboard")
+    public String loadDashboard(@RequestParam(required = false) String token, HttpServletRequest request) {
```
토큰 값을 파라미터보다는 request의 헤더 부분으로 받아오는 게 더 낫지 않을까 싶습니다.
1. 파라미터는 URL에 다 노출되기도 하고 (아마 JWT를 쿠키 말고 로컬 스토리지에 저장하시는 식으로 하신 것도 같은 이유 아닐까 생각해봅니다.)
2. 일반적인 규칙이 그렇다고 합니다.

jurio5 replied:
![1](https://github.com/user-attachments/assets/9f4c41d5-d169-4d5b-b487-b454dc246fa4)

쿼리 파라미터로 토큰이 전달되던 부분을 헤더 방식 받아올 수 있게 변경했습니다.

프론트에 구현한다. 라는 방식이 아직 감이 잘 안 잡히긴 합니다만, 한 번 해보겠습니다.

플러그인 같은 부분들은 

그, 주말에 Git을 사용하다가 익숙치가 않아서 여러가지 문제점에 봉착했는데 해결하는 과정에서 

이것 저것 만지다가 플러그인 까지 들어온걸로 보입니다.

혹은 제가 JWT관련 내용들을 긁어오다가 인스톨을 설치한 부분이 있는데 그 때 같이 들어왔을 수 있을 것 같습니다.

jurio5 replied:
+++ 

불필요한 파일들 전부 제거했습니다.

남은 파일들 중 필요없을 것 같은 파일이 남아있다면 말씀 부탁드립니다.

jurio5 replied:
+++

백엔드 , 프론트 서버 나눠서 리액트 작업 완료했습니다.

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,77 @@
+            jwtTokenProvider.invalidateToken(token);
+            return ResponseEntity.ok("로그아웃 되었습니다.");
+        }
+        return ResponseEntity.badRequest().body("유효하지 않은 토큰입니다.");
```
로그인 안 돼 있는데 로그아웃 요청 시 
(JWT 없으면 프론트에선 로그아웃 버튼이 애초에 안 보일텐데 로그아웃 요청을 보냈단건 정상적이지 않은 사용자일 가능성이 크므로)
1. 로그를 남기거나
2. 사용자를 정상 로그아웃의 경우와 다른 페이지로 이동시키거나

와 같이 정상 로그아웃 처리와 다른 처리를 할 게 아니라면
그냥 무조건 200을 반환해도 큰 문제는 없을 것 같아보여요. 아마도...?

jurio5 replied:
아, 네 좋은 것 같습니다. 유효성이 중복 되기도 했고, 이 부분도 고쳐서 커밋 했습니다.

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/admin/initializer/entity/AdminInitializer.java:
Admin 엔티티 클래스와 어떤 점이 다른 건가요?

hellonaeunkim replied:
똑같은 클래스이긴 한데, merge 할때 충돌 피하려고 나눴었습니다!
중복되는 내용이어서 Admin 엔티티 클래스로 통합시켰습니다.
커멘트 감사합니다!

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/admin/repository/AdminRepository.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.Optional;
+
+@Repository
```
JPA Repository의 경우 이 어노테이션이 없어도 Spring이 해당 인터페이스의 구현체를 자동으로 Bean으로 등록하는 걸로 알고 있어요. 그래서 이 어노테이션 제거하셔도 잘 작동할 겁니다.

jurio5 replied:
아, 네 Bean에 자동으로 등록이 되는건 처음 알았네요.

제거하고 테스트 했습니다.

감사합니다.

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/config/SecurityConfig.java:
```diff
@@ -0,0 +1,55 @@
+                .headers(headers -> headers.frameOptions(frameOptions -> frameOptions.disable()))
+                .sessionManagement(session -> session
+                        .sessionCreationPolicy(SessionCreationPolicy.STATELESS))
+                .authorizeHttpRequests(auth -> auth
```
.requestMatchers("/admin/**").hasAuthority("ROLE_ADMIN")
만 남기고 나머지 그냥 permitAll을 해도 될 것 같은데
나중에 유저 개인 페이지 같은 걸 염두에 두고 구분해놓으신 걸까요?

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/config/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,48 @@
+            token = request.getParameter("token");
+        }
+        
+        log.debug("수신된 토큰: {}", token);
```
파라미터에서 받아왔는데도 token 이 null이라면 바로 
``` java
response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "JWT is Missing");
return;
```
같이 바로 예외 줘버리면 굳이 컨트롤러까지 가지 않아도 될 것 같습니다.

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/admin/service/AdminService.java:
```diff
@@ -0,0 +1,80 @@
+        log.info("로그인 시도 - username: {}", username);
+        Admin admin = adminRepository.findByAdminUsername(username)
+                .orElseThrow(() -> {
+                    log.error("사용자를 찾을 수 없습니다.: {}", username);
```
저희 프로그램 자체에 해가 되는 치명적 오류 (해커로 추정되는 무수히 많은 로그인 시도 등) 같은 게 아니라면 
로그 레벨을 에러보다는
info나 warn 으로 써도 될 것 같다고 생각합니다.

로그는 사용자 (관리자나 고객)가 아니라 개발자가 보는 로그인데
단순히 누가 실수로 로그인 실패했다고 에러 로그 남아있으면 진짜 중요한 에러 로그를 못 볼 수도 있을 것 같아요.

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/config/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,117 @@
+    public boolean validateToken(String token) {
+        try {
+            if (blacklist.contains(token)) {
+                return false;
```
블랙 리스트라는게 밴 당한 사용자 같은 개념이 맞나요?
그렇다면 토큰값을 직접 비교하기보단 토큰 속의 username을 비교하는 게 맞다고 생각합니다.
토큰 값은 새로 로그인하거나 재발급 될 때마다 변하니까 블랙리스트에 넣은 의미가 무색해지지 않을까 싶어요.

InJunKangW replied:
다시 읽으니 무슨 의도인지 이해했습니다.
사용자가 로그아웃 시 해당 유저가 갖고 있던 JWT 를 다시 못 쓰게 한다는 것 같은데 맞나요?

InJunKangW reviewed (commented):
전체적으로 인증 인가 잘 작동해 보이는 코드라고 생각합니다.
몇몇 코멘트 남긴 부분들만 확인해주세요.
그리고 프론트용으로 html을 사용하시는 건 괜찮은데 백엔드 말고 프론트에 구현 부탁드립니다.

ps. 혹시 그리고 플러그인 같은 파일들이 굉장히 많이 생겼는데 혹시 의도하신 부분이실까요?
제가 잘 모르는 파일들이라...

InJunKangW commented on backend/src/main/java/com/example/nbe341team02/config/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    public void invalidateToken(String token) {
+        blacklist.add(token);
```
데이터가 적을 땐 괜찮은데 로그아웃한 횟수가 많으면 많을수록
Set의 size가 굉장히 커질 것 같아 염려됩니다.

dnzp75 replied:
보통 위와 같은 문제 때문에 blackList에는 [ key:value ] 형태로 [ 사용자 id : 토큰 값 ]  형태로 저장을 해서 사용을 합니다.

InJunKangW replied:
그럼 Map 형태로 바꾸시면 더 구현하기 편하실 것 같아요
[사용자(id든 username이든 식별만 가능하다면 뭐든지) : 리프레시 토큰 값] 형태의 맵을 만들고
로그인 시 토큰 발급하면서 리프레시 토큰 값을 맵에 넣고
로그아웃할 땐 맵에서 빼버리는 식으로..?

dnzp75 commented on backend/src/main/java/com/example/nbe341team02/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,77 @@
+    @GetMapping("/login")
+    public String showLoginPage() {
+        return "admin-login";
+    }
```
해당 api는 로그인 페이지를 보여주기 위한 api로 현재 사용되고 있나요?

jurio5 committed:
fix : parameter -> headers auth

dnzp75 commented on backend/src/main/java/com/example/nbe341team02/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,77 @@
+            }
+        }
+        return "redirect:/admin/login";
+    }
```
사용자가 아이디,비밀번호 입력 후 로그인 버튼 눌렀을 때 POST /login api가 요청이 되고, 해당 api가 성공적으로 수행되면 이 api가 요청이 될 것으로 보여요. 해당 api가 요청이 될 때에도 스프링에서 controller로 도달하기 전에 필터가 먼저 헤더로 들어온 토큰에 대해서 작성해주신 jwtAuthenticationFilter에서 먼저 검사를 하지 않을까요? 그렇게 생각해보면 여기서 토큰에 대한 검사를 하는 코드가 필요 없다고 생각이 드는데, 제가 흐름에 대해서 잘못 생각한 것이 있다면 말씀해주세요.

dnzp75 commented on backend/src/main/java/com/example/nbe341team02/admin/controller/AdminController.java:
```diff
@@ -0,0 +1,77 @@
+        }
+        return ResponseEntity.badRequest().body("유효하지 않은 토큰입니다.");
+    }
+}
```
여기서 마찬가지로 필터에서 먼저 토큰 유효성 검사를 이미 했을 것 같아요

dnzp75 commented on backend/src/main/java/com/example/nbe341team02/admin/service/AdminService.java:
```diff
@@ -0,0 +1,80 @@
+        return admin;
+    }
+
+   /* @PostConstruct
+    public void init() {
+        try {
+            // 기본 관리자 계정이 없는 경우에만 생성
+            if (adminRepository.findByAdminUsername("admin").isEmpty()) {
+                Admin admin = new Admin();
+                admin.setAdminUsername("admin");
+                admin.setAdminPassword(passwordEncoder.encode("admin123"));
+                admin.setAdminRole("ROLE_ADMIN");
+                adminRepository.save(admin);
+                log.info("기본 관리자 계정 생성");
+            }
+
+            if (adminRepository.findByAdminUsername("user").isEmpty()) {
+                Admin admin = new Admin();
+                admin.setAdminUsername("user");
+                admin.setAdminPassword(passwordEncoder.encode("user123"));
+                admin.setAdminRole("ROLE_USER");
+                adminRepository.save(admin);
+                log.info("기본 사용자 계정 생성");
+            }
+        } catch (Exception e) {
+            log.error("기본 계정 생성 실패", e);
+        }
+    }*/
```
사용되지 않는 주석처리된 코드는 PR 올려주실 때 지우고 올려주시는게 코드를 보는 입장에서 더 편리할 것 같다는 생각입니다!

dnzp75 commented on backend/src/main/java/com/example/nbe341team02/config/jwt/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,48 @@
+        // request 받으면 토큰 추출
+        String token = jwtTokenProvider.resolveToken(request);
+        
+        if (token == null) {
+            token = request.getParameter("token");
+        }
```
해당 검사는 토큰 관련 검사이기 때문에 resolveToken메서드 내에서 수행되는 것이 더 적합해 보이네요

jurio5 committed:
feat: JWT 인증 구현
- admin-login.html: 로그인 시 토큰 저장 및 대시보드 이동 로직 구현
- dashboard.html: 토큰 확인 로직 추가
- JwtAuthenticationFilter: 토큰 검증 로직 구현 & Add ignore

jurio5 committed:
Add : .gitignore

jurio5 committed:
.gitignore 추가 및 파일 정리

jurio5 committed:
Update .gitignore file

jurio5 committed:
Clean up project files and update .gitignore

jurio5 committed:
chore: update gitignore and remove conflict files

jurio5 committed:
chore: cleanup gitignore files Retry

jurio5 committed:
test

hellonaeunkim committed:
add : Application file

jurio5 committed:
feat : CORS guard React

jurio5 committed:
Add : React Admin login

jurio5 committed:
Add : run

jurio5 committed:
Add : run

jurio5 committed:
Fix : 비정상적인 로그아웃이 발생해서 토큰을 제거하고 로그인 페이지로 리다이렉트 되도록 변경

jurio5 committed:
Fix : 로그아웃 부분 수정

jurio5 committed:
feat : del login mapping

hellonaeunkim committed:
refacotr : 중복되는 AdminInitializer entity & repository Admin으로 통합

hellonaeunkim committed:
feat : blacklist에 저장된 token으로 접근 시도 시 오류 message 전송

jurio5 committed:
Fix : 백엔드, 프론트 분리 후 유저 권한 로그인 되던 부분 해결

jurio5 committed:
Fix : 백엔드 , 프론트 분리 후 인가 부분 해결

jurio5 committed:
Remove .idea/workspace.xml from git tracking

jurio5 committed:
Update .gitignore file

jurio5 committed:
Update .gitignore file

jurio5 committed:
Resolve merge conflicts

jurio5 committed:
Fix : application

jurio5 committed:
Merge main branch and resolve conflicts

jurio5 committed:
feat : 기존 중복되던 내용들 병합 처리

jurio5 committed:
Merge branch 'main' of https://github.com/prgrms-be-devcourse/NBE3-4-1-Team02

jurio5 committed:
feat : SecurityConfig

jurio5 committed:
Merge branch 'main' of https://github.com/prgrms-be-devcourse/NBE3-4-1-Team02

jurio5 committed:
Resolve merge conflicts

jurio5 committed:
Save changes before switching to main

jurio5 committed:
feat : securityconfig

notmybussiness committed:
Resolve conflicts and remove .DS_Store files
